src/Emogrifier/HtmlProcessor/HtmlPruner.php

```diff
@@ -41,4 +41,47 @@ public function removeInvisibleNodes()
 
         return $this;
     }
+
+    /**
+     * Removes classes that are no longer required (e.g. because there are no longer any CSS rules that reference them)
+     * from `class` attributes.
+     *
+     * Note that this does not inspect the CSS, but expects to be provided with a list of classes that are still in use.
+     *
+     * This method also has the (presumably beneficial) side-effect of minifying (removing superfluous whitespace from)
+     * `class` attributes.
+     *
+     * @param string[] $classesToKeep list of class names that should not be removed
+     *
+     * @return self fluent interface
+     */
+    public function removeRedundantClasses(array $classesToKeep)
+    {
+        $nodesWithClassAttribute = $this->xPath->query('//*[@class]');
+
+        if ($classesToKeep !== []) {
```

Comment:
Do we need this condition? Or is the second branch just a special case of the first one?

Reply:
> Or is the second branch just a special case of the first one?

It is indeed.  The first branch would achieve the same result, but do a whole lot of work for nothing - i.e. for every class attribute, doing a `preg_split`, computing the intersection with an empty array, testing that the result is empty (it of course always will be), then finally deducing that the attribute should be removed.

Reply:
I'm adding a comment to the code clarifying this.